ground/src/main/java/com/google/android/ground/ui/map/gms/features/FeatureManager.kt

```diff
@@ -0,0 +1,117 @@
+import timber.log.Timber
+
+/**
+ * Coordinates adding, removing, and updating of [Feature]s to the map. This abstracts access to
+ * both rendering of individual features, as well as clustering. This class is not thread-safe.
+ */
+class FeatureManager
+@Inject
+constructor(
+  @ApplicationContext private val context: Context,
+  private val pointRenderer: PointRenderer,
+  private val polygonRenderer: PolygonRenderer,
+  private val lineStringRenderer: LineStringRenderer
+) {
+  private val features = mutableSetOf<Feature>()
+  private val featuresByTag = mutableMapOf<Feature.Tag, Feature>()
+
+  private lateinit var map: GoogleMap
+  private lateinit var mapsItemManager: MapsItemManager
+  private lateinit var clusterManager: FeatureClusterManager
+  private lateinit var clusterRenderer: FeatureClusterRenderer
+
+  /**
+   * The camera's current zoom level. This must be set here since this impl can't access
+   * `map.cameraPosition` from off the main UI thread.
+   */
+  var zoom: Float
+    get() = clusterRenderer.zoom
+    set(value) {
+      clusterRenderer.zoom = value
+    }
+
+  /** Clears all managed state an binds to the provided [GoogleMap]. */
+  fun onMapReady(map: GoogleMap) {
+    features.clear()
+    featuresByTag.clear()
+    mapsItemManager = MapsItemManager(map, pointRenderer, polygonRenderer, lineStringRenderer)
+    clusterManager = FeatureClusterManager(context, map)
+    clusterRenderer = FeatureClusterRenderer(context, map, clusterManager, map.cameraPosition.zoom)
+    clusterRenderer.onClusterItemRendered = { mapsItemManager.setVisible(it, true) }
+    clusterRenderer.onClusterRendered = { mapsItemManager.setVisible(it, false) }
+    clusterManager.renderer = clusterRenderer
+    this.map = map
+  }
+
+  /**
+   * Updates the current set of features managed by the manager, adding and removing items from the
+   * map as needed to sync the map state with the provided collection.
+   */
+  fun setFeatures(updatedFeatures: Collection<Feature>) {
+    // remove stale
+    val removedOrChanged = features - updatedFeatures.toSet()
+    removedOrChanged.forEach(this::remove)
+    // add missing
+    val newOrChanged = updatedFeatures - features
+    newOrChanged.forEach { add(it) }
+    // cluster and update visibility
+    clusterManager.cluster()
+    Timber.v("${removedOrChanged.size} features removed, ${newOrChanged.size} added")
+  }
+
+  /**
+   * Returns the set of areas (polygon or multi-polygon features) which overlap with the specified
+   * coordinates.
+   */
+  fun getIntersectingPolygons(latLng: LatLng): Set<Feature> =
+    mapsItemManager.getIntersectingPolygonTags(latLng).mapNotNull { featuresByTag[it] }.toSet()
+
+  /**
+   * Adds a feature to the map, cluster, and to this class' internal index. Clusterable features are
+   * initialized as hidden so that the clusterer can determine whether they should be shown based on
+   * zoom level.
+   */
+  private fun add(feature: Feature) =
+    with(feature) {
```

Comment:
I don't think the `with` adds much value here, and it makes me have to pause and think ever so slightly more as a reader. this is exactly equivalent to just:

```kotlin
private fun add(feature: Feature) {
  features.add(feature)
  featuresByTag[tag] = feature
  if (feature.clusterable) clusterManager.addFeature(feature)
  mapsItemManager.put(feature, visible != feature.clusterable)
}
```

Which is slightly more verbose, but I like that I don't have to think about why we're using `with`. Unless I'm missing something here, I'm not sure the `this` binding value is totally there. In fact, I actually find it require a little bit more work to read implicit property access like an unqualified `clusterable`, but that might just be me.

To me, `with` makes most sense when you're doing a bunch of state mutations, i.e. you have lines like:

```kotlin
with(feature) {
  clusterable = false
  // ....
}.foo(//do some stuff with the mutated thing)
```

Reply:
I see two reasons why `with` might make this more readable:

* Everything we're doing here is with a `feature` instance; the `with` makes that context explicit at the top level of the function due to single-expression function syntax.
* We don't need to repeat `feature` multiple times, making the code more concise within that well-defined context.

Note in your above example, we'd need to quality feature one additional time for `tag`:

```kt
  private fun add(feature: Feature) {
    features.add(feature)
    featuresByTag[feature.tag] = feature
    if (feature.clusterable) clusterManager.addFeature(feature)
    mapsItemManager.put(feature, visible = !feature.clusterable)
  }
```

vs now:

```kt
  private fun add(feature: Feature) =
    with(feature) {
      features.add(this)
      featuresByTag[tag] = this
      if (clusterable) clusterManager.addFeature(this)
      mapsItemManager.put(this, visible = !clusterable)
    }
```

Change my mind! :)

Reply:
It's a fair point, I've just only ever seen `with` used mostly for the "configuration pattern" (as we use it in this PR for maps styling configuration in the renderers). 

BUT I don't *really* have a good argument against it in this case, so I think it's fine!

Reply:
I was looking around but couldn't find anything either way. The docs just have this to say:
https://kotlinlang.org/docs/scope-functions.html#with